crates/bevy_ecs/src/storage/sparse_set.rs

```diff
@@ -96,8 +98,8 @@ impl<I: SparseSetIndex, V> SparseArray<I, V> {
 pub struct ComponentSparseSet {
     dense: BlobVec,
     ticks: Vec<UnsafeCell<ComponentTicks>>,
-    entities: Vec<Entity>,
-    sparse: SparseArray<Entity, usize>,
+    entities: Vec<EntityId>,
```

Comment:
Won't this cause old data to be returned for despawned entities once the entity id is reused? Previously the entity would be of a different generation and thus couldn't return an old component for the previous generation of an entity.

Reply:
We actually never iterate over this. Instead using the entities stored in the Archetype for EntityFetch. This is purely used for bookkeeping.

Reply:
I think it isn't unsafe to return old components as a single `ComponentSparseSet` will always store components of the same type. It does however violate logical consistency.

Reply:
Then by the same logic, we should be validating any fetches against Tables with the same checks. In reality, these checks are non-trivial at this level, and are reliant on the machinery on top of the raw storage to ensure that those logical consistency guarantees are upheld.

We could (potentially) put this under a `#[cfg(debug_assertions)]` and store the full `Entitty` and use `debug_assert` to validate them.

Reply:
Added the debug assertions in, PTAL.

Reply:
:+1: